Mathlib/CategoryTheory/Types/Monomorphisms.lean

```diff
@@ -0,0 +1,128 @@
+namespace CategoryTheory
+
+open MorphismProperty Limits Types
+
+namespace Types
+
+instance : (monomorphisms (Type u)).IsStableUnderCobaseChange where
+  of_isPushout {X₁ X₂ X₃ X₄ t l r b} sq ht := by
+    simp only [monomorphisms.iff, mono_iff_injective] at ht ⊢
+    exact Limits.Types.pushoutCocone_inr_injective_of_isColimit sq.flip.isColimit ht
+
+lemma isStableUnderColimitsOfShape_monomorphisms_of_isFiltered
+    (J : Type u') [Category.{v'} J] [IsFiltered J] :
+    (monomorphisms (Type u)).IsStableUnderColimitsOfShape J := by
+  intro F₁ F₂ c₁ c₂ hc₁ hc₂ f hf
+  simp only [functorCategory, monomorphisms.iff, mono_iff_injective] at hf
+  let φ : c₁.pt ⟶ c₂.pt := hc₁.desc { ι := f ≫ c₂.ι, .. }
+  have hφ (j : J) : c₁.ι.app j ≫ φ = f.app j ≫ c₂.ι.app j := hc₁.fac _ j
+  replace hφ (j : J) := congr_fun (hφ j)
+  dsimp at hφ
+  change Mono φ
+  rw [mono_iff_injective]
+  intro x₁ y₁ h
+  obtain ⟨j, x₁, y₁, rfl, rfl⟩ : ∃ (j : J) (x₁' y₁' : F₁.obj j),
+      x₁ = c₁.ι.app j x₁' ∧ y₁ = c₁.ι.app j y₁' := by
+    obtain ⟨j, x₁, rfl⟩ := jointly_surjective_of_isColimit hc₁ x₁
+    obtain ⟨l, y₁, rfl⟩ := jointly_surjective_of_isColimit hc₁ y₁
+    exact ⟨_,  _, _, congr_fun (c₁.w (IsFiltered.leftToMax j l)).symm _,
+      congr_fun (c₁.w (IsFiltered.rightToMax j l)).symm _⟩
+  rw [hφ, hφ] at h
+  obtain ⟨k, α, hk⟩ := (FilteredColimit.isColimit_eq_iff' hc₂ _ _).1 h
+  simp only [← FunctorToTypes.naturality] at hk
+  rw [← c₁.w α, types_comp_apply, types_comp_apply, hf _ hk]
+
+section
+
+variable {J : Type u'} [LinearOrder J] [SuccOrder J] [OrderBot J] [WellFoundedLT J]
+
+namespace isStableUnderTransfiniteCompositionOfShape_monomorphisms
+
+variable {X Y : Type u} {f : X ⟶ Y}
+  (h : (monomorphisms (Type u)).TransfiniteCompositionOfShape J f)
+
+attribute [local instance] IsCofiltered.isConnected
+
+instance (j : J) : Mono (h.F.map (homOfLE bot_le : ⊥ ⟶ j)) := by
+  induction j using SuccOrder.limitRecOn with
+  | isMin j hj =>
+    obtain rfl := hj.eq_bot
+    exact inferInstanceAs (Mono (h.F.map (𝟙 _)))
+  | succ j hj hj' =>
+    have : Mono _ := h.map_mem j hj
+    rw [← homOfLE_comp bot_le (Order.le_succ j), Functor.map_comp]
+    infer_instance
+  | isSuccLimit j hj hj' =>
+    have : OrderBot (Set.Iio j) :=
+      { bot := ⟨⊥, Order.IsSuccLimit.bot_lt hj ⟩
+        bot_le _ := bot_le }
+    let φ : (Functor.const _).obj (h.F.obj ⊥) ⟶
+        (Set.principalSegIio j).monotone.functor ⋙ h.F :=
+      { app k := h.F.map (homOfLE bot_le)
+        naturality k k' hkk' := by
+          dsimp
+          rw [Category.id_comp, ← Functor.map_comp]
+          rfl }
+    have (k : Set.Iio j) : Mono (φ.app k) := hj' k.1 k.2
+    convert isStableUnderColimitsOfShape_monomorphisms_of_isFiltered _ _ _ _ _
+      (isColimitConstCocone (Set.Iio j) (h.F.obj ⊥))
+      (h.F.isColimitOfIsWellOrderContinuous j hj) φ
+        (fun _ ↦ monomorphisms.infer_property _)
+    apply (isColimitConstCocone (Set.Iio j) (h.F.obj ⊥)).hom_ext
+    intro j
+    rw [IsColimit.fac]
```

Comment:
This part appears in the lemma below as well, do you think it could be abstracted to a lemma about `W.IsStableUnderColimitsOfShape`?

Reply:
I have abstracted the argument in #24777.